### ISSUE TITLE
Description of Addressing (e asm.cpu) to be exposed in sdb.txt files

### ISSUE THREAD

**Issue by Maijin** [radare/radare2#8467]
During r2con several people expressed concern about the current way e asm.cpu (models of CPU) works, especially the fact that it requires to touch code for adding new CPU IO description (https://github.com/radare/radare2/blob/c0c0cba3398466661afed716f709f1fc404a752e/libr/syscall/ioports.c#L11) where it could go in a CPU files. (Parsing from (cmsis/SWD) via a script and put info in a sdb.txt for example)

Also it has been said that `e asm.cpu` in arm contains currently cortex-m and v8 where it should contains chips models like in AVR and therefore there should be a way to describe the ISA and the chipset. `e asm.cpu` is also said to be underused and could contain much more interesting information on the models to be discussed.

I have opened an HackMD for discussion https://hackmd.io/IwYwRgbBwJwCYFoBmBDJSEBYDsAmbCYK4hKAzAKxJiWa65JA

ping @killabytenow @queenp @dukebarman @Darkkey @brainstorm @xvilka

**Comment by queenp** [comment]
Quick opinions I have on this:
asm.cpu is not currently used in a consistent fashion. E.g.
- AVR, it refers to chipset (atmega1280, etc).
- For ARM it refers to choice between cortex (aka ARMv7) or v8 ARMv8 for the specific ARM instruction set supported.

The hardware ecology is enormous and I think it is useful as a general "cpu family"/"Instruction set" label.

From my PoV, what I would like is to have a way of importing existing hardware data description documents (e.g. https://github.com/posborne/cmsis-svd ) so that reversing stuff on particular chipsets is easier - if you see a ldr r0,[0x40013804] on an stm32l4x6 it can be flagged as a read on the CR2 register for the USART1 peripheral and not require continuously referring to the datasheet to look up memory mapped peripheral accesses to understand what they're doing.

Initially it wouldn't have to be able to do anything clever like emulation, just automatically recognising an address is listed in a system view description document.

The document format is specified [here](https://arm-software.github.io/CMSIS_5/SVD/html/index.html for general purpose further information. I understand they're not *always* consistent and often unofficial, but even a minimal python wrapper -> r2 specification converter would be surprisingly helpful (and it would leverage all the work that has been done elsewhere creating system descriptor files in the first place).

At the same time I understand that support for AVR has run in to some similar, related issues. I don't know so much about them, but I think it's important to create an integrated system with good separation of responsibilities between components so it's ergonomic for users to configure for whatever they're hacking on (like me, I really want to be able to do this with r2. :-) ).

**Comment by queenp** [comment]
Similar parseable documents also exist for AVR devices, in a somewhat different format (ATDF)

http://packs.download.atmel.com/

**Comment by Maijin** [comment]
Related to https://github.com/radare/radare2/issues/7746

**Comment by XVilka** [comment]
Check my previous unfinished PR for this. Will update it a bit later

**Comment by XVilka** [comment]
Basically I'm suggesting to use syscalls-like approach for both IO ports and MMIO mapped registers - like here https://github.com/radare/radare2/pull/8473

**Comment by brainstorm** [comment]
Hi XVilka! Sorry, but I don't see clearly how your PR organizes the different architectures "for outsiders" like myself? What will I see, as a user, when I run `e asm.cpu` on radare and how are those archs/address spaces organized behind the scenes?

**Comment by Maijin** [comment]
https://github.com/radare/radare2/pull/8473 from @XVilka

**Comment by Maijin** [comment]
See https://github.com/fluxchief/binaryninja_avr/blob/master/chips/iom16.py too

**Comment by Maijin** [comment]
Seems to define the name of the CPU, RAM_SIZE, ROM_SIZE, INTERRUPT_VECTOR_SIZE, IO_REGISTERS/EXTENDED_IO_REGISTERS

**Comment by radare** [comment]
ive merged the sysreg pr and rewrote it. so now the things are like this:

* rsyscall have a new api called r_syscall_sysreg() that accepts a type and a number to get the name of it
* this information is stored in Sdb
* overrides sysports information (the idea is that we will use sdb for io ports too)
* this file depends on arch+bits+cpu (OS is ignored)

**Comment by XVilka** [comment]
I will offer **0.1 BTC** to the one who will fix this issue. Initially, I wanted to use BountySource, but its support for bitcoins is a plain nightmare. So I will offer it directly.

**Comment by deepakchethan** [comment]
I am interested in doing this, and will start working on it post RSoC :)

**Comment by Maijin** [comment]
@deepakchethan if you are on telegram we can sync to discuss this task together with @XVilka

**Comment by deepakchethan** [comment]
@Maijin, Yes I am on telegram. I will ping you on it :)

**Comment by XVilka** [comment]
@deepakchethan so, are you up for this task?

**Comment by deepakchethan** [comment]
Yes, i'm going through a lot of changes in my IRL and will get back to radare2 contributions by the end of this week. I am sorry about the delay.

**Comment by XVilka** [comment]
@deepakchethan any updates regarding this one?

**Comment by 08A** [comment]
I am working on r_syscall_sysreg. Where do you want to store the info on io, reg, mem?
If we store this info in the sdb file it is interpreted as an array.

**Comment by 08A** [comment]
We could have a collection of sdb file for every specific archi, and the user could load the specific file if he want.
example: archi-info-chiprandom.sdb
with something like this inside
```
mem.09090.name=truc
mem.09090.comment='fait un truc'

reg.rtruc.comment='est vraiment super utile'
```

**Comment by Maijin** [comment]
Wait - there is a lot to do here, don't rush things out, don't forget that this issue  is a project for GSOC, you will have to come up with a calendar/planning you'll discuss with the mentor of that project! But yes one of the idea is to have files with the configuration per CPUs instead of hardcoded configuration.

Understanding and making r_syscall_sysreg working is a first step - check how that's working now and you can improve it on the model of the e asm.describe and a simple sdb structure.

**Comment by ret2libc** [comment]
This issue has been moved from radareorg/radare2 to radareorg/ideas as we are trying to clean our backlog and this issue has probably been created a long while ago. This is an effort to help contributors understand what are the actionable items they can work on, prioritize issues better and help users find active/duplicated issues more easily. If this is not an enhancement/improvement/general idea but a bug, feel free to ask for re-transfer to main repo. Thanks for your understanding and contribution with this issue.
